tracing-benchmarks/src/jmh/java/com/palantir/tracing/TracingBenchmark.java

```diff
@@ -0,0 +1,152 @@
+@Fork(1)
+@Threads(16)
+@SuppressWarnings("checkstyle:hideutilityclassconstructor")
+public class TracingBenchmark {
+
+    private static final Callable<Object> NOOP_CALLABLE = () -> null;
+
+    private static final Callable<Object> undecidedCallable = Tracers.wrapWithNewTrace("test",
+            Observability.UNDECIDED, NOOP_CALLABLE);
+
+    private static final Callable<Object> sampledCallable = Tracers.wrapWithNewTrace("test",
+            Observability.SAMPLE, NOOP_CALLABLE);
+
+    private static final Callable<Object> unsampledCallable = Tracers.wrapWithNewTrace("test",
+            Observability.DO_NOT_SAMPLE, NOOP_CALLABLE);
+
+    @Setup
+    public final void before() {
+        Tracer.setSampler(new RandomSampler(0.1f));
+    }
+
+    @TearDown
+    public final void after() {
+        Tracer.setSampler(AlwaysSampler.INSTANCE);
+    }
+
+    @Benchmark
+    public static Object raw() throws Exception {
+        return NOOP_CALLABLE.call();
+    }
+
+    @Benchmark
+    public static Object undecidedCallable() throws Exception {
+        return undecidedCallable.call();
+    }
+
+    @Benchmark
+    public static Object sampledCallable() throws Exception {
+        return sampledCallable.call();
+    }
+
+    @Benchmark
+    public static Object unsampledCallable() throws Exception {
+        return unsampledCallable.call();
+    }
+
+    @Benchmark
+    public static void unsampledTracerSpan(Blackhole blackhole) throws Exception {
+        Tracer.initTrace(Observability.DO_NOT_SAMPLE, Tracers.randomId());
+        for (int i = 0; i < 100; i++) {
+            Callable<Object> callable = () -> {
+                Tracer.startSpan("span");
+                try {
+                    return NOOP_CALLABLE.call();
+                } finally {
+                    Tracer.fastCompleteSpan();
+                }
+            };
+            blackhole.consume(callable.call());
+        }
+    }
+
+    @Benchmark
+    public static void undecidedTracerSpan(Blackhole blackhole) throws Exception {
+        Tracer.initTrace(Observability.UNDECIDED, Tracers.randomId());
+        for (int i = 0; i < 100; i++) {
+            Callable<Object> callable = () -> {
+                Tracer.startSpan("span");
+                try {
+                    return NOOP_CALLABLE.call();
+                } finally {
+                    Tracer.fastCompleteSpan();
+                }
+            };
+            blackhole.consume(callable.call());
+        }
+    }
+
+    @Benchmark
+    public static void sampledTracerSpan(Blackhole blackhole) throws Exception {
+        Tracer.initTrace(Observability.SAMPLE, Tracers.randomId());
+        for (int i = 0; i < 100; i++) {
+            Callable<Object> callable = () -> {
+                Tracer.startSpan("span");
+                try {
+                    return NOOP_CALLABLE.call();
+                } finally {
+                    Tracer.fastCompleteSpan();
+                }
```

Comment:
Should precompute the callable wrappers.

Reply:
done